packages/ci/src/lib/monorepo/list-projects.ts

```diff
@@ -0,0 +1,121 @@
+import { glob } from 'glob';
+import { join } from 'node:path';
+import type { Logger, Settings } from '../models';
+import { detectMonorepoTool } from './detect-tool';
+import { getToolHandler } from './handlers';
+import { listPackages } from './packages';
+import type { MonorepoHandlerOptions, ProjectConfig } from './tools';
+
+export async function listMonorepoProjects(
+  settings: Settings,
+): Promise<ProjectConfig[]> {
+  if (!settings.monorepo) {
+    throw new Error('Monorepo mode not enabled');
+  }
+
+  const logger = settings.logger;
+
+  const options = createMonorepoHandlerOptions(settings);
+
+  const tool =
+    settings.monorepo === true
+      ? await detectMonorepoTool(options)
+      : settings.monorepo;
+  if (settings.monorepo === true) {
+    if (tool) {
+      logger.info(`Auto-detected monorepo tool ${tool}`);
+    } else {
+      logger.info("Couldn't auto-detect any supported monorepo tool");
+    }
+  } else {
+    logger.info(`Using monorepo tool "${tool}" from inputs`);
+  }
+
+  if (tool) {
+    const handler = getToolHandler(tool);
+    const projects = await handler.listProjects(options);
+    logger.info(`Found ${projects.length} projects in ${tool} monorepo`);
+    logger.debug(`Projects: ${projects.map(({ name }) => name).join(', ')}`);
+    return projects;
+  }
+
+  if (settings.projects) {
+    return listProjectsByGlobs({
+      patterns: settings.projects,
+      cwd: options.cwd,
+      bin: settings.bin,
+      logger,
+    });
+  }
+
+  return listProjectsByNpmPackages({
+    cwd: options.cwd,
+    bin: settings.bin,
+    logger,
+  });
+}
+
+function createMonorepoHandlerOptions(
+  settings: Settings,
+): MonorepoHandlerOptions {
+  return {
+    task: settings.task,
+    cwd: settings.directory,
+    ...(!settings.silent && {
+      observer: {
+        onStdout: stdout => {
+          console.info(stdout);
+        },
+        onStderr: stderr => {
+          console.warn(stderr);
+        },
```

Comment:
This configuration could be a new option `verbose` and we would still be able to add custom logs outside.

Reply:
Good point, although I'm wondering if it shouldn't be on by default :thinking: Instead of _opting in_ to printing the logs with `verbose` (default is off by convention), we would _opt out_ of those logs using a `silent` flag (inverted logic).

The GitHub Actions toolkit exports its own versions of our `executeProcess` ([`exec`](https://github.com/actions/toolkit/blob/main/packages/exec/src/exec.ts#L7-L31) and [`getExecOutput`](https://github.com/actions/toolkit/blob/main/packages/exec/src/exec.ts#L33-L90)), where their [`silent` flag is `false` by default](https://github.com/actions/toolkit/blob/main/packages/exec/src/interfaces.ts#L12-L13).

But this is maybe beyond the scope of this PR.

Reply:
Moved into new issue:
- #843